mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -172,6 +174,36 @@ class BlockMatrix(
     assert(cols <= nCols, s"The number of columns $cols is more than claimed $nCols.")
   }
 
+  def validate(): Unit = {
+    logDebug("Validating BlockMatrix...")
+    // check if the matrix is larger than the claimed dimensions
+    estimateDim()
+    logDebug("BlockMatrix dimensions are okay...")
+
+    // Check if there are multiple MatrixBlocks with the same index.
+    val indexCounts = blockInfo.countByKey().foreach { case (key, cnt) =>
+      if (cnt > 1) {
+        throw new SparkException(s"There are MatrixBlocks with duplicate indices. Please remove " +
+          s"blocks with duplicate indices. You may call reduceByKey on the underlying RDD and " +
+          s"sum the duplicates. You may convert the matrices to Breeze before summing them up.")
+      }
+    }
+    logDebug("MatrixBlock indices are okay...")
+    // Check if each MatrixBlock (except edges) has the dimensions rowsPerBlock x colsPerBlock
+    // The first tuple is the index and the second tuple is the dimensions of the MatrixBlock
+    blockInfo.foreach { case ((blockRowIndex, blockColIndex), (m, n)) =>
+      if (m != rowsPerBlock || n != colsPerBlock) {
```

Comment:
This is not correct. Checking `m` should be separate from checking `n`. `m` is valid if 

~~~
(blockRowIndex < numRowBlocks - 1 && m == rowsPerBlock) || (blockRowIndex == numRowBlocks - 1 && m > 0 && m <= rowsPerBlock)
~~~